src/verilog/verilog_preprocessor.h

```diff
@@ -60,45 +73,84 @@ class verilog_preprocessort:public preprocessort
 
   std::list<conditionalt> conditionals;
 
-  // for include
+  class vector_token_sourcet : public verilog_preprocessor_token_sourcet
+  {
+  public:
+    vector_token_sourcet(const std::vector<tokent> &_tokens)
+      : tokens(_tokens), pos(tokens.begin())
+    {
+    }
+
+  protected:
+    const std::vector<tokent> &tokens;
+    std::vector<tokent>::const_iterator pos;
+    void get_token_from_stream() override;
+  };
 
-  class filet
+  // for include and for `define
+  class contextt
   {
   protected:
-    bool close;
+    bool deallocate_in;
     std::istream *in;
     std::string filename;
 
   public:
-    verilog_preprocessor_tokenizert tokenizer;
+    verilog_preprocessor_token_sourcet *tokenizer;
+
+    // for `define with parameters
+    define_argumentst define_arguments;
+
+    contextt(bool _deallocate_in, std::istream *_in, std::string _filename)
+      : deallocate_in(_deallocate_in),
+        in(_in),
+        filename(std::move(_filename)),
+        tokenizer(new verilog_preprocessor_tokenizert(*in))
+    {
+    }
 
-    filet(bool _close, std::istream *_in, std::string _filename)
-      : close(_close), in(_in), filename(std::move(_filename)), tokenizer(*in)
+    explicit contextt(const std::vector<tokent> &tokens)
+      : deallocate_in(false),
+        in(nullptr),
+        filename(),
+        tokenizer(new vector_token_sourcet(tokens))
```

Comment:
Why does this need to be a pointer?

Reply:
I suppose you mean the `vector_token_sourcet`, not the input stream.
This is the C version of a variant type. One of the constructors makes a `vector_token_sourcet`, the other a `verilog_preprocessor_tokenizert`. With C++17, this could be a `std::variant`.